packages/react-instantsearch-core/src/components/InstantSearch.tsx

```diff
@@ -137,64 +130,63 @@ class InstantSearch extends Component<Props, State> {
     stalledSearchDelay: PropTypes.number,
   };
 
-  isControlled: boolean;
-  isUnmounting: boolean;
-  aisManager: InstantSearchManager;
-
-  constructor(props: Props) {
-    super(props);
-    this.isControlled = Boolean(props.searchState);
-    const initialState = this.isControlled ? props.searchState : {};
-    this.isUnmounting = false;
-
-    this.aisManager = createInstantSearchManager({
-      indexName: props.indexName,
-      searchClient: props.searchClient,
-      initialState,
-      resultsState: props.resultsState,
-      stalledSearchDelay: props.stalledSearchDelay,
-    });
-
-    this.state = {
+  static getDerivedStateFromProps(
+    nextProps: Props,
+    prevState: State
+  ): Partial<State> {
+    const nextIsControlled = isControlled(nextProps);
+
+    if (!prevState.isControlled && nextIsControlled) {
+      throw new Error(
+        "You can't switch <InstantSearch> from being uncontrolled to controlled"
+      );
+    }
+
+    if (prevState.isControlled && !nextIsControlled) {
+      throw new Error(
+        "You can't switch <InstantSearch> from being controlled to uncontrolled"
+      );
+    }
+
+    return {
+      isControlled: nextIsControlled,
```

Comment:
We can't switch from controlled to uncontrolled, it means that we can compute the value once for all. It was the case inside the previous implementation: the value was computed inside the constructor that's it. The check was performed inside cWRP. We can't do it there anymore but we can move the error inside cDU. It will throw an error anyway, sooner or later it doesn't matter.